File: Python/sudoku.py
import numpy as np
import time
import random
import sys
from solver import Sudoku

FILENAME = "../Data/show_sudoku.csv"
SHOW_GUI = True
NO_PUZZLES = 1
SIZE = 9
 
# Read puzzles from csv-file
def read_puzzles(filename, no_puzzles):
    puzzles = np.full((no_puzzles, SIZE * SIZE), None, dtype=object)
    solutions = np.full((no_puzzles, SIZE * SIZE), None, dtype=object)

    with open(filename, 'r') as csv_file:
        all_sudokus = csv_file.read().splitlines()
        if len(all_sudokus) + 1 < NO_PUZZLES:
            raise ValueError("There are not enough puzzles in file")
    
    sudokus = random.sample(all_sudokus[1:], k=no_puzzles)
    for i, line in enumerate(sudokus):
        puzzle, solution = line.split(",")
        for j, q_s in enumerate(zip(puzzle, solution)):
            q, s = q_s
            puzzles[i, j] = int(q) if q != '0' else None
            solutions[i, j] = int(s) if s != '0' else None

    puzzles = puzzles.reshape((-1, SIZE, SIZE))
    solutions = solutions.reshape((-1, SIZE, SIZE))

    return puzzles, solutions    

def solve_sudokus(algorithm, show=False):
    start_time = time.time()
    puzzles, solutions = read_puzzles(FILENAME, NO_PUZZLES)                 # Get sudokus
    download_end_time = time.time()
    if show: print(f"It took {round(download_end_time - start_time, 4)} seconds to download {NO_PUZZLES} sudokus")

    # Solve each sudoku
    solved_sudokus = 0
    total_tries, total_empty_cells = 0, 0
    for puzzle, solution in zip(puzzles, solutions):
        sudoku = Sudoku(puzzle, solution)
        tries, empty_cells = sudoku.solve(algorithm, SHOW_GUI)
        
        if sudoku.correct_solution():
            solved_sudokus += 1
            total_tries += tries
            total_empty_cells += empty_cells

    solve_end_time = time.time()
    if show: print(f"{algorithm} solved {round(solved_sudokus / NO_PUZZLES, 2)*100}% of {NO_PUZZLES} sudokus in {round(solve_end_time - download_end_time, 4)} seconds")
    try:
        if show: print(f"{algorithm} averaged {round(total_tries / total_empty_cells, 2)} tries per unsolved cell")
    except ZeroDivisionError:
        if show: print("No sudoku solved")

    return round(solve_end_time - download_end_time, 4)

def main():
    try: 
        algorithm = sys.argv[1]
        return solve_sudokus(algorithm)
    except IndexError:
        print("Choose an algorithm")
        return
    
if __name__ == '__main__':
    main()

File: C/sudoku.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#include "functions.h"

int main() {
    int (*quizzes)[SIZE][SIZE] = malloc(NO_PUZZLES * sizeof(*quizzes));
    int (*solutions)[SIZE][SIZE] = malloc(NO_PUZZLES * sizeof(*solutions));
    int solvedSudokus = 0;

    // Load all sudokus
    clock_t startTime = clock();
    int count = ReadQuizzes(FILENAME, NO_PUZZLES, quizzes, solutions);
    clock_t downloadEndTime = clock();
    printf("It took %f seconds to load %d puzzles.\n", (double)(downloadEndTime - startTime) / CLOCKS_PER_SEC, count);

    // Solve all sudokus
    for (int i = 0; i < count; i++) {
        if (SolveSudoku(&quizzes[i], &solutions[i]) == 1) {
            solvedSudokus++;
        } 
    }  

    clock_t solve_end_time = clock();
    printf("Method solved %f%% of %d sudokus in %f seconds.\n",
           (double)solvedSudokus / NO_PUZZLES * 100, NO_PUZZLES,
           (double)(solve_end_time - downloadEndTime) / CLOCKS_PER_SEC);

    free(quizzes);
    free(solutions);
    return 1;
}

int ReadQuizzes(const char* filename, int puzzles, int (*sudoku)[SIZE][SIZE], int (*solution)[SIZE][SIZE]) {
    // Handle missing / corrupted file
    FILE *file = fopen(filename, "r");
    if (!file) {
        printf("Error opening file\n");
        return 0;
    }

    char line[1024];
    int puzzleCount = 0;
    int firstLine = 1;

    // Read all sudokus
    while (fgets(line, sizeof(line), file) && puzzleCount < puzzles) {
        if (firstLine) {
            firstLine = 0;
            continue;
        }

        char *inputSudoku = strtok(line, ",");
        char *inputSolution = strtok(NULL, "\n");

        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                sudoku[puzzleCount][i][j] = inputSudoku[i * SIZE + j] - '0';
                solution[puzzleCount][i][j] = inputSolution[i * SIZE + j] - '0';
            }
        }
        puzzleCount++;
    }

    fclose(file);
    return puzzleCount;
}

void PrintSudoku(int (*sudoku)[SIZE][SIZE]) {
    for (int row = 0; row < SIZE; row++) {
        for (int col = 0; col < SIZE; col++) {
            printf("%2d", (*sudoku)[row][col]);
        }
        printf("\n");
    }
}